File: 62.Kahn'sAlgo.java
/* ----------------------------------------------------------------------------  */
/*   ( The Authentic JS/JAVA CodeBuff )
 ___ _                      _              _ 
 | _ ) |_  __ _ _ _ __ _ __| |_ __ ____ _ (_)
 | _ \ ' \/ _` | '_/ _` / _` \ V  V / _` || |
 |___/_||_\__,_|_| \__,_\__,_|\_/\_/\__,_|/ |
                                        |__/ 
 */
/* --------------------------------------------------------------------------   */
/*    Youtube: https://youtube.com/@code-with-Bharadwaj                        */
/*    Github : https://github.com/Manu577228                                  */
/*    Portfolio : https://manu-bharadwaj-portfolio.vercel.app/portfolio       */
/* -----------------------------------------------------------------------  */

import java.io.*;
import java.util.*;

public class KahnsAlgorithm {

    // Function implementing Kahn's Algorithm for Topological Sort
    static List<Integer> kahnTopologicalSort(int vertices, int[][] edges) {
        // Step 1: Create adjacency list
        Map<Integer, List<Integer>> graph = new HashMap<>();
        for (int i = 0; i < vertices; i++) graph.put(i, new ArrayList<>());
        for (int[] edge : edges) graph.get(edge[0]).add(edge[1]);
        // Example: if edges are {5,0}, {4,0} => graph = {5:[0], 4:[0]}

        // Step 2: Compute indegree for each vertex
        int[] indegree = new int[vertices];
        for (int u : graph.keySet()) {
            for (int v : graph.get(u)) indegree[v]++;
        }
        // indegree[v] = number of incoming edges to vertex v

        // Step 3: Initialize queue with vertices having indegree = 0
        Queue<Integer> q = new LinkedList<>();
        for (int i = 0; i < vertices; i++) {
            if (indegree[i] == 0) q.offer(i);
        }
        // These are starting points (no dependencies)

        // Step 4: Process queue
        List<Integer> topoOrder = new ArrayList<>();
        while (!q.isEmpty()) {
            int node = q.poll();          // Take one vertex with indegree 0
            topoOrder.add(node);          // Add it to topological order

            // Decrease indegree of all its adjacent vertices
            for (int neighbor : graph.get(node)) {
                indegree[neighbor]--;
                // If indegree becomes 0, add to queue
                if (indegree[neighbor] == 0) q.offer(neighbor);
            }
        }

        // Step 5: Check if topological sort is possible (no cycles)
        if (topoOrder.size() == vertices) return topoOrder;
        return new ArrayList<>(); // Graph has a cycle
    }

    public static void main(String[] args) throws Exception {
        int V = 6;
        int[][] E = {
            {5, 2}, {5, 0}, {4, 0},
            {4, 1}, {2, 3}, {3, 1}
        };

        List<Integer> order = kahnTopologicalSort(V, E);
        System.out.println("Topological Order: " + order);
    }
}

File: 62.Kahn'sAlgo.py
# In a Directed Acyclic Graph, topological sort arranges the vertices such that for every directed 
# edge u → v, vertex u comes before v in the ordering.
# Kahn’s Algorithm uses in-degree (the number of edges directed into a node) 
# and a queue to maintain processing order.

# Explanation:

# Compute the in-degree for each vertex.

# Push all vertices with in-degree = 0 into a queue.

# While the queue is not empty:

# Remove a vertex from the queue and add it to the topological order.

# For each of its neighbors, reduce their in-degree by 1.

# If any neighbor’s in-degree becomes 0, add it to the queue.

# If all vertices are processed, you get a valid topological order. Otherwise, the graph has a cycle.

from collections import defaultdict, deque

def kahn_topological_sort(vertices, edges):
    graph = defaultdict(list)
    for u, v in edges:
        graph[u].append(v)

    indegree = [0] * vertices
    for u in graph:
        for v in graph[u]:
            indegree[v] += 1

    q = deque()
    for i in range(vertices):
        if indegree[i] == 0:
            q.append(i)

    topo_order = []
    while q:
        node = q.popleft()
        topo_order.append(node)

        for neighbor in graph[node]:
            indegree[neighbor] -= 1
            if indegree[neighbor] == 0:
                q.append(neighbor)

    if len(topo_order) == vertices:
        return topo_order
    else:
        return []
    
V = 6
E = [(5, 2), (5, 0), (4, 0), (4, 1), (2, 3), (3, 1)]

order = kahn_topological_sort(V, E)
print("Topological Order:", order)
